### PR TITLE
Handle signals

### PR DESCRIPTION
**Pull request by roi-codefresh**

### CODE REVIEW

File: venona/pkg/server/server.go [hunk 1]
```diff
@@ -15,31 +15,105 @@
...
 
 	"github.com/codefresh-io/go/venona/pkg/logger"
 	"github.com/gin-gonic/gin"
 )
 
 var (
-	errAlreadyStarted = errors.New("Already started")
+	errAlreadyStarted    = errors.New("Already started")
+	errAlreadyStopped    = errors.New("Already stopped")
+	errMustProvideLogger = errors.New("Must provide logger")
```
Comment (olegsu) [inline venona/pkg/server/server.go:29]: "Logger is required"

File: venona/pkg/server/server.go [hunk 2]
```diff
@@ -15,31 +15,105 @@
...
+		Port   string
+		Logger logger.Logger
+		Mode   string
+	}
+
 	// Server is an HTTP server that expose API
 	Server struct {
-		Port    string
 		Logger  logger.Logger
+		EventsC chan Event
```
Comment (olegsu) [inline venona/pkg/server/server.go:46]: name it `EventsChan` please

File: venona/pkg/server/server.go [hunk 3]
```diff
@@ -15,31 +15,105 @@
...
+	}
+}
+
+func (s Server) shutdown() {
+	ctx := context.Background()
+	err := s.srv.Shutdown(ctx)
+	if err != nil {
+		s.Logger.Error("failed to gracefully terminate server, cause: ", err)
+	}
+	s.started = false
```
Comment (olegsu) [inline venona/pkg/server/server.go:118]: s.running instead of s.started

File: venona/cmd/start.go [hunk 4]
```diff
@@ -133,9 +143,54 @@ func run(options startOptions) {
...
+		Mode:   serverMode,
+	})
+	dieOnError(err)
+
+	go handleSignals(context.Background(), &agent, &server, log)
+
 	dieOnError(server.Start())
 }
+
+func handleSignals(ctx context.Context, a *agent.Agent, s *server.Server, log logger.Logger) {
```
Comment (olegsu) [inline venona/cmd/start.go:162]: do not use the server or agent here to avoid race condition, instead pass the channel

File: venona/cmd/start.go [hunk 5]
```diff
@@ -133,9 +143,54 @@ func run(options startOptions) {
...
+	handleSignal(sigChan, syscall.SIGTERM, syscall.SIGINT) // sent by k8s
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case sig := <-sigChan:
+			switch sig {
+			case syscall.SIGTERM, syscall.SIGINT:
+				go func() {
```
Comment (olegsu) [inline venona/cmd/start.go:176]: Not sure why we need run another routine here

File: venona/go.mod [hunk 6]
```diff
@@ -14,6 +14,7 @@ require (
 	github.com/spf13/viper v1.7.0
 	github.com/stretchr/objx v0.2.0 // indirect
 	github.com/stretchr/testify v1.5.1
+	github.com/vektra/mockery v1.1.2 // indirect
```
Comment (olegsu) [inline venona/go.mod:17]: this line should be deleted, after that run `go mod tidy`

File: venona/cmd/start.go [hunk 7]
```diff
@@ -66,7 +75,7 @@ func init() {
 	viper.SetDefault("codefresh-host", defaultCodefreshHost)
 	viper.SetDefault("port", "8080")
 
-	startCmd.Flags().BoolVar(&startCmdOptions.verbose, "verbose", true, "Show more logs")
+	startCmd.Flags().BoolVar(&startCmdOptions.verbose, "verbose", false, "Show more logs")
```
Comment (olegsu) [inline venona/cmd/start.go:78]: Now the default verbose is false, i think we need to update some log calls from `Debug` to `Info` across the app

File: venona/cmd/start.go (hunk 4 above)
Comment (roi-codefresh) [reply venona/cmd/start.go:162]: right, will do

File: venona/cmd/start.go (hunk 5 above)
Comment (roi-codefresh) [reply venona/cmd/start.go:176]: running another routine to allow the "force-stop" option, don't want to block the handling of signals and wait for server to finish.

File: venona/cmd/start.go (hunk 7 above)
Comment (roi-codefresh) [reply venona/cmd/start.go:78]: Changed a few important calls from Debug to Info

**Pull request merged**